Sources/PackageMetadata/PackageMetadata.swift

```diff
@@ -0,0 +1,147 @@
+                                                                                           readmeURL: $0.readmeURL) }
+                    if packages.isEmpty, let error = error {
+                        // If the search result is empty and we had a previous error, emit it now.
+                        return callback(.failure(error))
+                    } else {
+                        return callback(.success(packages))
+                    }
+                } catch {
+                    return callback(.failure(error))
+                }
+            }
+        }
+
+        let fetchStandalonePackageByURL = {
+            guard let url = URL(string: query) else {
+                return search(nil)
+            }
+
+            do {
+                try withTemporaryDirectory(removeTreeOnDeinit: true) {
+                    let tempPath = $0.appending(component: url.lastPathComponent)
+                    do {
+                        let repositorySpecifier = RepositorySpecifier(url: url)
+                        try self.repositoryProvider.fetch(repository: repositorySpecifier, to: tempPath, progressHandler: nil)
+                        if self.repositoryProvider.isValidDirectory(tempPath), let repository = try self.repositoryProvider.open(repository: repositorySpecifier, at: tempPath) as? GitRepository {
+                            let branches = try repository.getBranches()
+                            let versions = try repository.getTags().compactMap { Version($0) }
+                            let package = Package(identity: .init(url: url),
+                                                  location: url.absoluteString,
+                                                  branches: branches,
+                                                  versions: versions,
+                                                  readmeURL: self.guessReadMeURL(baseURL: url, defaultBranch: try repository.getDefaultBranch()))
+                            return callback(.success([package]))
+                        }
+                    } catch {
+                        return search(error)
+                    }
+                }
+            } catch {
+                return search(error)
+            }
+        }
+
+        if isRegistryIdentity {
+            return self.registryClient.getPackageMetadata(package: identity, observabilityScope: observabilityScope, callbackQueue: DispatchQueue.sharedConcurrent) { result in
+                do {
+                    let metadata = try result.get()
+                    let readmeURL: URL?
+                    if let alternateURL = metadata.alternateLocations?.first {
+                        // FIXME: This is pretty crude, we should let the registry metadata provide the value instead.
+                        readmeURL = guessReadMeURL(baseURL: alternateURL, defaultBranch: "main")
```

Comment:
We probably want this in the registry metadata instead

Reply:
We could perhaps add `readmeURL` as a reserved key to the [API spec](https://github.com/apple/swift-package-manager/blob/main/Documentation/Registry.md#4621-reserved-metadata-keys) and have this code use it?

Reply:
Yah, I think that makes sense. Probably in a follow-up PR.